File: tetris_net/server_go/ws.go
package main

// ws.go — WebSocket(RFC 6455) 서버 쪽을 직접 구현한다. 외부 패키지를 쓰지 않는다.
//
// 왜 직접 쓰나: 이 문서가 가르치려는 것 중 하나가 "웹소켓은 대단한 프로토콜이 아니다"라는
// 사실이다. 핸드셰이크는 SHA-1 한 번이고, 프레임은 2바이트 머리 + 길이 + (마스크) + 본문이다.
// 아래 200줄이 그 전부다. 게다가 의존성이 0이면 `go build` 하나로 단일 바이너리가 나온다.
//
// 다루지 않는 것: 확장(permessage-deflate), 서브프로토콜 협상, 클라이언트 쪽 Dial.
// 이 게임에 필요 없다. 필요해지면 그때 넣으면 된다.

import (
	"bufio"
	"crypto/rand"
	"crypto/sha1"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// 프레임 종류 (RFC 6455 §5.2). 상위 비트(0x8)가 켜져 있으면 제어 프레임이다.
const (
	OpCont   byte = 0x0
	OpText   byte = 0x1
	OpBinary byte = 0x2
	OpClose  byte = 0x8
	OpPing   byte = 0x9
	OpPong   byte = 0xA
)

// RFC 6455 §1.3 이 못 박아 둔 마법 문자열. 이 값 때문에 캐시나 프록시가
// 핸드셰이크 응답을 "그냥 HTTP 응답"으로 착각해 재사용하지 못한다.
const wsGUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"

// AcceptKey — 클라이언트가 보낸 Sec-WebSocket-Key 에서 응답 헤더 값을 만든다.
// base64(SHA1(key + GUID)). 암호학적 의미는 없고, "나는 웹소켓을 아는 서버다"라는 증표다.
func AcceptKey(key string) string {
	h := sha1.New()
	io.WriteString(h, key)
	io.WriteString(h, wsGUID)
	return base64.StdEncoding.EncodeToString(h.Sum(nil))
}

// Conn — 업그레이드가 끝난 연결 하나.
type Conn struct {
	c          net.Conn
	br         *bufio.Reader
	wmu        sync.Mutex // 쓰기는 프레임 단위로 원자적이어야 한다. 여러 고루틴이 같이 쓴다
	MaxPayload int64      // 한 메시지의 상한. 이 게임의 최대 메시지는 1KB 도 안 된다
	Client     bool       // 클라이언트 쪽이면 방향이 뒤집힌다 (보낼 때 마스킹, 받을 때 요구 안 함)
	closed     bool
}

func NewConn(c net.Conn) *Conn { return newConn(c, bufio.NewReader(c)) }
func newConn(c net.Conn, br *bufio.Reader) *Conn {
	return &Conn{c: c, br: br, MaxPayload: 1 << 20}
}

// Upgrade — HTTP 요청 하나를 WebSocket 연결로 바꾼다.
// http.Hijacker 로 소켓을 통째로 빼앗아 오는 게 핵심이다. 그 뒤로는 net/http 가
// 이 연결에 손을 대지 않으므로 우리가 프레임을 직접 읽고 쓸 수 있다.
func Upgrade(w http.ResponseWriter, r *http.Request) (*Conn, error) {
	if !strings.EqualFold(r.Header.Get("Upgrade"), "websocket") {
		return nil, errors.New("Upgrade 헤더가 websocket 이 아니다")
	}
	if !strings.Contains(strings.ToLower(r.Header.Get("Connection")), "upgrade") {
		return nil, errors.New("Connection 헤더에 upgrade 가 없다")
	}
	if r.Header.Get("Sec-WebSocket-Version") != "13" {
		return nil, errors.New("Sec-WebSocket-Version 이 13 이 아니다")
	}
	key := r.Header.Get("Sec-WebSocket-Key")
	if key == "" {
		return nil, errors.New("Sec-WebSocket-Key 가 없다")
	}
	hj, ok := w.(http.Hijacker)
	if !ok {
		return nil, errors.New("이 ResponseWriter 는 Hijack 을 지원하지 않는다")
	}
	c, brw, err := hj.Hijack()
	if err != nil {
		return nil, err
	}
	// 101 을 손으로 쓴다. 여기서부터는 HTTP 가 아니다.
	_, err = fmt.Fprintf(brw, "HTTP/1.1 101 Switching Protocols\r\n"+
		"Upgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Accept: %s\r\n\r\n", AcceptKey(key))
	if err == nil {
		err = brw.Flush()
	}
	if err != nil {
		c.Close()
		return nil, err
	}
	// 하이재킹으로 받은 bufio.Reader 를 그대로 써야 한다.
	// 클라이언트가 101 을 기다리지 않고 첫 프레임을 붙여 보냈다면 그게 여기 들어 있다.
	return newConn(c, brw.Reader), nil
}

// ── 프레임 쓰기 ──
// 머리와 본문을 **한 번에** 쓴다. 두 번 나눠 쓰면 작은 프레임이 두 개의 TCP 세그먼트로
// 쪼개져 나가고, 상대가 헤더만 읽고 멈추는 상황이 생긴다(테스트가 이걸 잡는다).
func writeRaw(w io.Writer, fin bool, op byte, p []byte, mask bool) error {
	n := 2
	l := len(p)
	var head [14]byte
	head[0] = op & 0x0f
	if fin {
		head[0] |= 0x80
	}
	switch {
	case l < 126:
		head[1] = byte(l)
	case l <= 0xffff:
		head[1] = 126
		binary.BigEndian.PutUint16(head[2:], uint16(l))
		n = 4
	default:
		head[1] = 127
		binary.BigEndian.PutUint64(head[2:], uint64(l))
		n = 10
	}
	var key [4]byte
	if mask {
		head[1] |= 0x80
		if _, err := rand.Read(key[:]); err != nil {
			return err
		}
		copy(head[n:], key[:])
		n += 4
	}
	buf := make([]byte, n+l)
	copy(buf, head[:n])
	copy(buf[n:], p)
	if mask {
		for i := 0; i < l; i++ {
			buf[n+i] ^= key[i&3]
		}
	}
	_, err := w.Write(buf)
	return err
}

// WriteFrame — FIN 이 켜진 프레임 하나. 테스트에서 "클라이언트인 척" 할 때도 쓴다.
func WriteFrame(w io.Writer, op byte, p []byte, mask bool) error {
	return writeRaw(w, true, op, p, mask)
}

// 서버가 내보내는 프레임은 절대 마스킹하지 않는다 (RFC 6455 §5.1).
func (c *Conn) write(op byte, p []byte) error {
	c.wmu.Lock()
	defer c.wmu.Unlock()
	if c.closed {
		return net.ErrClosed
	}
	return writeRaw(c.c, true, op, p, c.Client)
}
func (c *Conn) WriteText(p []byte) error { return c.write(OpText, p) }
func (c *Conn) Close() error {
	c.wmu.Lock()
	c.closed = true
	c.wmu.Unlock()
	return c.c.Close()
}

// ── 프레임 읽기 ──
func (c *Conn) readFrame() (fin bool, op byte, payload []byte, err error) {
	var h [2]byte
	if _, err = io.ReadFull(c.br, h[:]); err != nil {
		return
	}
	fin = h[0]&0x80 != 0
	if h[0]&0x70 != 0 {
		err = errors.New("RSV 비트가 켜져 있다 (확장을 협상한 적이 없다)")
		return
	}
	op = h[0] & 0x0f
	masked := h[1]&0x80 != 0
	if !masked && !c.Client {
		// 클라이언트→서버 프레임은 반드시 마스킹돼야 한다. 이건 보안이 아니라
		// 중간의 낡은 캐시 프록시가 본문을 HTTP 로 오해하지 못하게 하는 장치다.
		err = errors.New("마스킹되지 않은 클라이언트 프레임")
		return
	}
	if masked && c.Client {
		err = errors.New("서버가 마스킹된 프레임을 보냈다")
		return
	}
	l := int64(h[1] & 0x7f)
	switch l {
	case 126:
		var e [2]byte
		if _, err = io.ReadFull(c.br, e[:]); err != nil {
			return
		}
		l = int64(binary.BigEndian.Uint16(e[:]))
	case 127:
		var e [8]byte
		if _, err = io.ReadFull(c.br, e[:]); err != nil {
			return
		}
		u := binary.BigEndian.Uint64(e[:])
		if u > 1<<62 {
			err = errors.New("길이가 말이 안 된다")
			return
		}
		l = int64(u)
	}
	if l > c.MaxPayload {
		err = fmt.Errorf("프레임이 상한을 넘었다 (%d > %d)", l, c.MaxPayload)
		return
	}
	var key [4]byte
	if masked {
		if _, err = io.ReadFull(c.br, key[:]); err != nil {
			return
		}
	}
	payload = make([]byte, l)
	if _, err = io.ReadFull(c.br, payload); err != nil {
		return
	}
	if masked {
		for i := range payload { // 마스크 해제 = 4바이트 키로 XOR
			payload[i] ^= key[i&3]
		}
	}
	return
}

// ReadMessage — 조각난 프레임을 이어 붙여 메시지 하나를 돌려준다.
// 제어 프레임(핑/퐁/닫기)은 여기서 삼키고 애플리케이션에 올리지 않는다.
func (c *Conn) ReadMessage() (byte, []byte, error) {
	var msgOp byte
	var buf []byte
	for {
		fin, op, p, err := c.readFrame()
		if err != nil {
			return 0, nil, err
		}
		if op&0x8 != 0 { // 제어 프레임
			if !fin || len(p) > 125 {
				return 0, nil, errors.New("제어 프레임은 조각날 수 없고 125바이트를 넘을 수 없다")
			}
			switch op {
			case OpPing:
				if err := c.write(OpPong, p); err != nil {
					return 0, nil, err
				}
			case OpClose:
				// 닫기는 인사를 돌려주고 끝낸다. 상대가 이미 사라졌을 수 있으니
				// 여기서 영원히 매달리지 않도록 짧은 마감을 건다.
				c.c.SetWriteDeadline(time.Now().Add(500 * time.Millisecond))
				c.write(OpClose, p)
				c.c.SetWriteDeadline(time.Time{})
				return 0, nil, io.EOF
			case OpPong: // 살아 있다는 신호. 할 일 없음
			}
			continue
		}
		if op == OpCont {
			if msgOp == 0 {
				return 0, nil, errors.New("이어 붙일 메시지가 없다")
			}
			buf = append(buf, p...)
		} else {
			if msgOp != 0 {
				return 0, nil, errors.New("앞 메시지가 끝나지 않았다")
			}
			msgOp = op
			buf = p
		}
		if int64(len(buf)) > c.MaxPayload {
			return 0, nil, errors.New("조각 합계가 상한을 넘었다")
		}
		if fin {
			return msgOp, buf, nil
		}
	}
}

// ── 클라이언트 쪽 ──
// 서버만 있으면 테스트를 못 쓴다. Dial 은 40줄이면 되고, 운영 도구로도 쓸모가 있다.
// Client 가 켜지면 방향이 뒤집힌다 — 보낼 때 마스킹하고, 받을 때 마스킹을 요구하지 않는다.
func Dial(rawurl string) (*Conn, error) {
	u, err := url.Parse(rawurl)
	if err != nil {
		return nil, err
	}
	host := u.Host
	if !strings.Contains(host, ":") {
		host += ":80"
	}
	c, err := net.Dial("tcp", host)
	if err != nil {
		return nil, err
	}
	var nonce [16]byte
	if _, err := rand.Read(nonce[:]); err != nil {
		c.Close()
		return nil, err
	}
	key := base64.StdEncoding.EncodeToString(nonce[:])
	path := u.RequestURI()
	if path == "" {
		path = "/"
	}
	req := fmt.Sprintf("GET %s HTTP/1.1\r\nHost: %s\r\nUpgrade: websocket\r\nConnection: Upgrade\r\n"+
		"Sec-WebSocket-Key: %s\r\nSec-WebSocket-Version: 13\r\n\r\n", path, u.Host, key)
	if _, err := io.WriteString(c, req); err != nil {
		c.Close()
		return nil, err
	}
	br := bufio.NewReader(c)
	resp, err := http.ReadResponse(br, nil)
	if err != nil {
		c.Close()
		return nil, err
	}
	resp.Body.Close()
	if resp.StatusCode != 101 {
		c.Close()
		return nil, fmt.Errorf("업그레이드 거절: %s", resp.Status)
	}
	if got := resp.Header.Get("Sec-WebSocket-Accept"); got != AcceptKey(key) {
		c.Close()
		return nil, errors.New("Sec-WebSocket-Accept 가 맞지 않는다")
	}
	conn := newConn(c, br)
	conn.Client = true
	return conn, nil
}

File: tetris_net/server_py/ws.py
# -*- coding: utf-8 -*-
"""ws.py — WebSocket(RFC 6455) 을 표준 라이브러리만으로 구현한다.

`pip install` 이 필요 없다는 게 이 파일의 존재 이유다. 파이썬만 있으면 서버가 뜬다.
Go 의 ws.go 와 같은 일을 하고, 같은 테스트(RFC §1.3 예제·길이 3갈래·마스킹)를 통과한다.

다루지 않는 것: 확장(permessage-deflate), 서브프로토콜 협상, TLS(그건 앞단 nginx 몫이다).
"""
import asyncio
import base64
import hashlib
import json
import secrets
import struct
import urllib.parse

OP_CONT, OP_TEXT, OP_BINARY, OP_CLOSE, OP_PING, OP_PONG = 0x0, 0x1, 0x2, 0x8, 0x9, 0xA

# RFC 6455 §1.3 이 못 박아 둔 마법 문자열. 한 글자만 틀려도 브라우저가 연결을 거부한다.
GUID = '258EAFA5-E914-47DA-95CA-C5AB0DC85B11'

MAX_PAYLOAD = 1 << 20
_LIMIT = 1 << 20


def accept_key(key: str) -> str:
    """Sec-WebSocket-Key → Sec-WebSocket-Accept. base64(SHA1(key + GUID))."""
    return base64.b64encode(hashlib.sha1((key + GUID).encode()).digest()).decode()


def build_frame(op: int, payload: bytes, mask: bool = False, fin: bool = True) -> bytes:
    """프레임 한 개를 통째로 만든다. 머리와 본문을 나눠 쓰지 않는 게 중요하다 —
    작은 프레임이 두 세그먼트로 쪼개져 나가면 상대가 머리만 읽고 멈추는 일이 생긴다."""
    b0 = (0x80 if fin else 0) | (op & 0x0F)
    n = len(payload)
    mbit = 0x80 if mask else 0
    if n < 126:
        head = struct.pack('!BB', b0, mbit | n)
    elif n <= 0xFFFF:
        head = struct.pack('!BBH', b0, mbit | 126, n)
    else:
        head = struct.pack('!BBQ', b0, mbit | 127, n)
    if not mask:
        return head + payload
    key = secrets.token_bytes(4)
    return head + key + bytes(b ^ key[i & 3] for i, b in enumerate(payload))


class WSConn:
    """업그레이드가 끝난 연결 하나.

    client=True 면 방향이 뒤집힌다 — 보낼 때 마스킹하고, 받을 때 마스킹을 요구하지 않는다.
    (RFC 6455 §5.1: 클라이언트→서버는 반드시 마스킹, 서버→클라이언트는 절대 마스킹 금지)
    """

    def __init__(self, reader, writer, client=False, max_payload=MAX_PAYLOAD):
        self.reader, self.writer, self.client = reader, writer, client
        self.max_payload = max_payload
        self._wlock = asyncio.Lock()
        self._closed = False

    # ── 읽기 ──
    async def _read_frame(self):
        h = await self.reader.readexactly(2)
        if h[0] & 0x70:
            raise ValueError('RSV 비트가 켜져 있다 (확장을 협상한 적이 없다)')
        fin, op = bool(h[0] & 0x80), h[0] & 0x0F
        masked = bool(h[1] & 0x80)
        if not masked and not self.client:
            # 마스킹 요구는 보안이 아니라, 낡은 캐시 프록시가 본문을 HTTP 로
            # 오해하지 못하게 하려는 장치다.
            raise ValueError('마스킹되지 않은 클라이언트 프레임')
        if masked and self.client:
            raise ValueError('서버가 마스킹된 프레임을 보냈다')
        n = h[1] & 0x7F
        if n == 126:
            n = struct.unpack('!H', await self.reader.readexactly(2))[0]
        elif n == 127:
            n = struct.unpack('!Q', await self.reader.readexactly(8))[0]
        if n > self.max_payload:
            raise ValueError('프레임이 상한을 넘었다 (%d)' % n)
        key = await self.reader.readexactly(4) if masked else None
        data = await self.reader.readexactly(n) if n else b''
        if masked:
            data = bytes(b ^ key[i & 3] for i, b in enumerate(data))
        return fin, op, data

    async def recv(self):
        """조각난 프레임을 이어 붙여 메시지 하나를 돌려준다. 끝났으면 None.
        제어 프레임(핑/퐁/닫기)은 여기서 삼키고 애플리케이션에 올리지 않는다."""
        msg_op, buf = None, bytearray()
        while True:
            try:
                fin, op, data = await self._read_frame()
            except (asyncio.IncompleteReadError, ConnectionError, ValueError, OSError):
                return None
            if op & 0x8:                       # 제어 프레임
                if not fin or len(data) > 125:
                    return None
                if op == OP_PING:
                    await self.send(OP_PONG, data)
                elif op == OP_CLOSE:
                    await self.send(OP_CLOSE, data)
                    return None
                continue                        # 퐁은 살아 있다는 신호. 할 일 없음
            if op == OP_CONT:
                if msg_op is None:
                    return None
            else:
                if msg_op is not None:
                    return None
                msg_op = op
            buf += data
            if len(buf) > self.max_payload:
                return None
            if fin:
                return msg_op, bytes(buf)

    async def recv_json(self):
        m = await self.recv()
        if m is None:
            return None
        op, data = m
        if op != OP_TEXT:
            return {}
        try:
            return json.loads(data.decode('utf-8'))
        except (ValueError, UnicodeDecodeError):
            return {}

    # ── 쓰기 ──
    async def send(self, op, data: bytes):
        if self._closed:
            return
        async with self._wlock:                 # 프레임은 원자적으로 나가야 한다
            try:
                self.writer.write(build_frame(op, data, mask=self.client))
                await self.writer.drain()
            except (ConnectionError, OSError):
                self._closed = True

    async def send_text(self, s: str):
        await self.send(OP_TEXT, s.encode('utf-8'))

    async def send_json(self, obj):
        # ensure_ascii=False 로 보내야 한글 이름이 \uXXXX 로 부풀지 않는다.
        await self.send_text(json.dumps(obj, ensure_ascii=False, separators=(',', ':')))

    async def close(self):
        if self._closed:
            return
        self._closed = True
        try:
            self.writer.close()
            await self.writer.wait_closed()
        except (ConnectionError, OSError):
            pass


# ── 핸드셰이크 ──
def _parse_head(raw: bytes):
    lines = raw.decode('latin1').split('\r\n')
    parts = lines[0].split(' ')
    if len(parts) < 2:
        return None, '', {}
    headers = {}
    for line in lines[1:]:
        if ':' in line:
            k, v = line.split(':', 1)
            headers[k.strip().lower()] = v.strip()
    return parts[0], parts[1], headers


async def serve(handler, host, port, http=None):
    """웹소켓 서버를 연다. handler(conn, path) 가 연결 하나의 일생을 맡는다.
    업그레이드가 아닌 요청은 http(path) 에 넘긴다 — (상태, 콘텐츠타입, 본문) 또는 None."""

    async def on_client(reader, writer):
        try:
            raw = await asyncio.wait_for(reader.readuntil(b'\r\n\r\n'), 10)
        except (asyncio.TimeoutError, asyncio.IncompleteReadError, OSError,
                asyncio.LimitOverrunError):
            writer.close()
            return
        method, path, headers = _parse_head(raw)
        up = headers.get('upgrade', '').lower() == 'websocket'
        key = headers.get('sec-websocket-key', '')
        if method != 'GET' or not up or not key or headers.get('sec-websocket-version') != '13':
            status, ctype, body = (http(path) if http else None) or \
                (404, 'text/plain; charset=utf-8', '없는 경로다\n'.encode())
            writer.write(('HTTP/1.1 %d OK\r\nContent-Type: %s\r\nContent-Length: %d\r\n'
                          'Connection: close\r\n\r\n' % (status, ctype, len(body))).encode('latin1'))
            writer.write(body)
            try:
                await writer.drain()
            except OSError:
                pass
            writer.close()
            return
        writer.write(('HTTP/1.1 101 Switching Protocols\r\nUpgrade: websocket\r\n'
                      'Connection: Upgrade\r\nSec-WebSocket-Accept: %s\r\n\r\n'
                      % accept_key(key)).encode('latin1'))
        await writer.drain()
        conn = WSConn(reader, writer)
        try:
            await handler(conn, path)
        finally:
            await conn.close()

    return await asyncio.start_server(on_client, host, port, limit=_LIMIT)


async def connect(url: str) -> WSConn:
    """클라이언트 쪽. 테스트와 운영 도구가 쓴다 — 40줄이면 된다."""
    u = urllib.parse.urlparse(url)
    reader, writer = await asyncio.open_connection(u.hostname, u.port or 80, limit=_LIMIT)
    key = base64.b64encode(secrets.token_bytes(16)).decode()
    path = u.path or '/'
    if u.query:
        path += '?' + u.query
    writer.write(('GET %s HTTP/1.1\r\nHost: %s\r\nUpgrade: websocket\r\nConnection: Upgrade\r\n'
                  'Sec-WebSocket-Key: %s\r\nSec-WebSocket-Version: 13\r\n\r\n'
                  % (path, u.netloc, key)).encode('latin1'))
    await writer.drain()
    raw = await reader.readuntil(b'\r\n\r\n')
    lines = raw.decode('latin1').split('\r\n')
    if '101' not in lines[0]:
        writer.close()
        raise ConnectionError('업그레이드 거절: %s' % lines[0])
    _, _, headers = _parse_head(raw)
    if headers.get('sec-websocket-accept') != accept_key(key):
        writer.close()
        raise ConnectionError('Sec-WebSocket-Accept 가 맞지 않는다')
    return WSConn(reader, writer, client=True)
